apps/algolia/lib/algolia/api.ex

```diff
@@ -6,45 +6,59 @@ defmodule Algolia.Api do
   require Logger
   use RepoCache, ttl: :timer.hours(12)
 
-  defstruct [:host, :index, :action, :body]
+  defstruct [:host, :index, :action, :body, :query_params]
+
+  @type action :: :post | :get
 
   @http_pool Application.get_env(:algolia, :http_pool)
 
   @type t :: %__MODULE__{
           host: String.t() | nil,
           index: String.t() | nil,
           action: String.t() | nil,
-          body: String.t() | nil
+          body: String.t() | nil,
+          query_params: map | nil
         }
 
-  @spec post(t) :: {:ok, HTTPoison.Response.t()} | {:error, HTTPoison.Error.t() | :bad_config}
-  def post(opts, config \\ nil)
+  @spec action(action, t) ::
+          {:ok, HTTPoison.Response.t()} | {:error, HTTPoison.Error.t() | :bad_config}
+  def action(action, opts, config \\ nil)
 
-  def post(%__MODULE__{} = opts, nil) do
-    post(opts, Config.config())
+  def action(action, %__MODULE__{} = opts, nil) do
+    action(action, opts, Config.config())
   end
 
-  def post(%__MODULE__{} = opts, %Config{write: <<_::binary>>, app_id: <<_::binary>>} = config) do
-    do_post(opts, config)
+  def action(
+        action,
+        %__MODULE__{} = opts,
+        %Config{write: <<_::binary>>, app_id: <<_::binary>>} = config
+      ) do
+    do_action(action, opts, config)
   end
 
-  def post(%__MODULE__{}, %Config{} = config) do
+  def action(_action, %__MODULE__{}, %Config{} = config) do
     _ = Logger.warn("module=#{__MODULE__} missing Algolia config keys: #{inspect(config)}")
     {:error, :bad_config}
   end
 
-  defp do_post(%__MODULE__{index: index, action: action, body: body} = opts, %Config{} = config)
-       when is_binary(index) and is_binary(action) and is_binary(body) do
+  defp do_action(
+         action,
+         %__MODULE__{index: index, action: opts_action, body: body} = opts,
+         %Config{} = config
+       )
+       when is_binary(index) and is_binary(opts_action) and is_binary(body) do
     hackney =
       opts
       |> hackney_opts()
       |> Keyword.put(:pool, @http_pool)
 
     send_post_request = fn {body, config} ->
+      query_param_string = generate_query_param_string(opts)
```

Comment:
nit: I feel like this should be done inside the `generate_url` function, as it's part of generating the URL!